base/applications/rapps/appdb.cpp

```diff
@@ -141,74 +141,124 @@ CAppDB::UpdateAvailable()
     EnumerateFiles();
 }
 
-VOID
-CAppDB::UpdateInstalled()
+static inline HKEY
+GetRootKeyInfo(UINT Index, REGSAM &RegSam)
 {
-    // Remove all old entries
-    ClearList(m_Installed);
+    C_ASSERT(_countof(g_RootKeyEnum) == _countof(g_RegSamEnum));
+    if (Index < _countof(g_RootKeyEnum))
+    {
+        RegSam = g_RegSamEnum[Index];
+        return g_RootKeyEnum[Index];
+    }
+    return NULL;
+}
 
-    int LoopKeys = 2;
+HKEY
+CAppDB::EnumInstalledRootKey(UINT Index, REGSAM &RegSam)
+{
+    // Loop for through all combinations.
+    // Note that HKEY_CURRENT_USER\Software does not have a redirect
+    // https://docs.microsoft.com/en-us/windows/win32/winprog64/shared-registry-keys#redirected-shared-and-reflected-keys-under-wow64
+    if (Index < (IsSystem64Bit() ? 3 : 2))
+        return GetRootKeyInfo(Index, RegSam);
+    else
+        return NULL;
+}
 
-    if (IsSystem64Bit())
+CInstalledApplicationInfo *
+CAppDB::CreateInstalledAppByRegistryKey(LPCWSTR KeyName, HKEY hKeyParent, UINT KeyIndex)
+{
+    CRegKey hSubKey;
+    if (hSubKey.Open(hKeyParent, KeyName, KEY_READ) == ERROR_SUCCESS)
     {
-        // loop for all 3 combination.
-        // note that HKEY_CURRENT_USER\Software don't have a redirect
-        // https://docs.microsoft.com/en-us/windows/win32/winprog64/shared-registry-keys#redirected-shared-and-reflected-keys-under-wow64
-        LoopKeys = 3;
+        DWORD value, size;
+
+        size = sizeof(DWORD);
+        if (!RegQueryValueExW(hSubKey, L"SystemComponent", NULL, NULL, (LPBYTE)&value, &size) && value == 1)
+        {
+            // Ignore system components
+            return NULL;
+        }
+
+        size = 0;
+        BOOL bIsUpdate = !RegQueryValueExW(hSubKey, L"ParentKeyName", NULL, NULL, NULL, &size);
+
+        AppsCategories cat = bIsUpdate ? ENUM_UPDATES : ENUM_INSTALLED_APPLICATIONS;
+        CInstalledApplicationInfo *pInfo;
+        pInfo = new CInstalledApplicationInfo(hSubKey.Detach(), KeyName, cat, KeyIndex);
+        if (pInfo && pInfo->Valid())
+        {
+            return pInfo;
+        }
+        delete pInfo;
     }
+    return NULL;
+}
 
-    for (int keyIndex = 0; keyIndex < LoopKeys; keyIndex++)
+CInstalledApplicationInfo *
+CAppDB::EnumerateRegistry(CAtlList<CAppInfo *> *List, LPCWSTR SearchOnly)
+{
+    ATLASSERT(List || SearchOnly);
```

Comment:
- This method is private to the class
- It does not make sense to call it with both set to NULL because you will get nothing returned, it should not crash however.